### PR TITLE
PW-6778 Support gift cards using Order API

### PR DESCRIPTION
**Pull request by zaiddreakh**
## Summary

- This PR is for SFRA to support gift cards using the web component and the Order API to create a partial payments order
- For now it only supports 1 gift card issuer (brand), and only 1 gift card per order

## Tested scenarios

**Fixed issue**:

### CODE REVIEW

File: src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderGenericComponent.js [hunk 1]
```diff
@@ -135,13 +160,22 @@ module.exports.renderGenericComponent = async function renderGenericComponent()
...
   setCheckoutConfiguration(store.checkout.options);
   setInstallments(store.checkout.options.amount);
   setAmazonPayConfig(store.checkout.paymentMethodsResponse);
   document.querySelector('#paymentMethodsList').innerHTML = '';
 
+  store.checkout.paymentMethodsResponse.paymentMethods.some((pm) => {
+    if (pm.type === 'giftcard') {
+      renderGiftCard(pm);
+    }
+    return null;
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderGenericComponent.js:176]: do we need to return anything?

File: src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderPaymentMethod.js [hunk 2]
```diff
@@ -57,7 +58,7 @@ function getPaymentMethodID(isStored, paymentMethod) {
   }
   if (paymentMethod.brand) {
     // gift cards all share the same type. Brand is used to differentiate between them
-    return `${paymentMethod.type}_${paymentMethod.brand}`;
+    return 'giftcard';
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderPaymentMethod.js:61]: cards also have a brand, does this not mess things up for card component?

File: src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderPaymentMethod.js [hunk 3]
```diff
@@ -160,14 +162,17 @@ module.exports.renderPaymentMethod = function renderPaymentMethod(
   const imagePath = getImagePath(options);
   const liContents = getListContents({ ...options, imagePath, description });
 
-  li.innerHTML = liContents;
-  li.classList.add('paymentMethod');
-
+  if (rerender) {
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderPaymentMethod.js:165]:
```suggestion
  const li = rerender ? document.querySelector(`#rb_${paymentMethodID}`).closest('li'): document.createElement('li');
```

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/Adyen.js [hunk 4]
```diff
@@ -77,6 +77,27 @@ function getExternalPlatformVersion() {
   return EXTERNAL_PLATFORM_VERSION;
 }
 
+/**
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/Adyen.js:80]: Can we move all these to above the `getExternalPlatformVersion` function?

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/cancelPartialPaymentOrder.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    const cancelOrderRequest = {
+      merchantAccount: AdyenConfigs.getAdyenMerchantAccount(),
+      order: splitPaymentsOrder,
+    };
+
+    const response = adyenCheckout.doCancelPartialPaymentOrderCall(
+      cancelOrderRequest,
+    );
+
+    if (response.resultCode && response.resultCode === 'Received') {
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/cancelPartialPaymentOrder.js:21]: first part of if clause is superfluous

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/cancelPartialPaymentOrder.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+          }
+        });
+      });
+      session.privacy.giftCardResponse = null;
+    }
+
+    res.json(response);
+    return next();
+  } catch (error) {
+    return next();
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/cancelPartialPaymentOrder.js:36]: if we're not handling any errors we might as well move the return statement to under the try/catch block

Comment (wboereboom) [reply src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/cancelPartialPaymentOrder.js:36]: We could consider logging if that's not already handled by the function calling the service

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/checkBalance.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+const BasketMgr = require('dw/order/BasketMgr');
+const AdyenHelper = require('*/cartridge/scripts/util/adyenHelper');
+const adyenCheckout = require('*/cartridge/scripts/adyenCheckout');
+const AdyenConfigs = require('*/cartridge/scripts/util/adyenConfigs');
+
+function callCheckBalance(req, res, next) {
+  try {
+    const currentBasket = BasketMgr.getCurrentBasket();
+    const request = JSON.parse(req.body);
+    let paymentMethod;
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/checkBalance.js:11]: do we let this remain undefined if paymentMethod isn't on the request?

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/checkBalance.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        ).value,
+      },
+      reference: currentBasket.getUUID(),
+      paymentMethod,
+    };
+
+    const response = adyenCheckout.doCheckBalanceCall(checkBalanceRequest);
+
+    res.json(response);
+    return next();
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/checkBalance.js:31]: we can move the next return to outside the try catch

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/partialPayment.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      }); // entire response exceeds string length
+    });
+
+    const remainingAmount = new Money(
+      response.order.remainingAmount.value,
+      response.order.remainingAmount.currency,
+    ).divide(100);
+    response.remainingAmountFormatted = remainingAmount.toFormattedString();
+    res.json(response);
+    return next();
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/partialPayment.js:36]: we can move + deduplicate the return statement again here

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/partialPayment.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+    const partialPaymentRequest = {
+      merchantAccount: AdyenConfigs.getAdyenMerchantAccount(),
+      amount,
+      reference: 'partialPaymentRef',
+      paymentMethod,
+      order: splitPaymentsOrder,
+    };
+
+    const response = adyenCheckout.doPaymentsCall(0, 0, partialPaymentRequest);
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/partialPayment.js:21]: I understand why these are `0`, but only after checking input args for the doPaymentsCall function. Maybe we can add some comments and/or input `null` or `undefined` or something

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/paymentFromComponent.js [hunk 11]
```diff
@@ -71,6 +78,30 @@ function paymentFromComponent(req, res, next) {
     }
   }
 
+  // Check if gift card was used
+  if (session.privacy.giftCardResponse) {
+    let giftcardPM;
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/paymentFromComponent.js:83]: I think we can remove this and declare the const at line 87

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/splitPayments.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+function addMinutes(date, minutes) {
+  return new Date(date.getTime() + minutes * 60000);
+}
+
+function createSplitPaymentsOrder(req, res, next) {
+  try {
+    const currentBasket = BasketMgr.getCurrentBasket();
+
+    let date = new Date();
+    date = addMinutes(date, 30);
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/splitPayments.js:16]: Maybe we can make this a const so it's easy to override?

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/splitPayments.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+      reference: currentBasket.getUUID(),
+      expiresAt: date.toISOString(),
+    };
+
+    const response = adyenCheckout.doCreateSplitPaymentOrderCall(
+      splitPaymentsRequest,
+    );
+
+    res.json(response);
+    return next();
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/splitPayments.js:35]: again we can dedoop

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/checkout_services/placeOrder.js [hunk 14]
```diff
@@ -139,6 +142,26 @@ function placeOrder(req, res, next) {
 
     // Handles payment authorization
     var handlePaymentResult = adyenHelpers.handlePayments(order);
+
+    //Check if gift card was used
+    if(session.privacy.giftCardResponse) {
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/checkout_services/placeOrder.js:147]: Is this where you create the second payment instrument after order creation?

File: src/cartridges/int_adyen_overlay/cartridge/scripts/adyenCheckout.js [hunk 15]
```diff
@@ -82,13 +82,19 @@ function createPaymentRequest(args) {
...
-      value: myAmount,
-    };
+    // Add split payments order if applicable
+    if (paymentInstrument.custom.adyenSplitPaymentsOrder) {
+      paymentRequest.order = JSON.parse(paymentInstrument.custom.adyenSplitPaymentsOrder).splitPaymentsOrder;
+      paymentRequest.amount = JSON.parse(paymentInstrument.custom.adyenSplitPaymentsOrder).remainingAmount;
+    } else {
+      const myAmount = AdyenHelper.getCurrencyValueForApi(
+          paymentInstrument.paymentTransaction.amount,
+      ).getValueOrNull(); // args.Amount * 100;
```
Comment (wboereboom) [inline src/cartridges/int_adyen_overlay/cartridge/scripts/adyenCheckout.js:92]: is this comment still needed?

File: src/cartridges/int_adyen_overlay/cartridge/scripts/util/adyenHelper.js [hunk 16]
```diff
@@ -550,10 +551,15 @@ var adyenHelperObj = {
 
   // saves the payment details in the paymentInstrument's custom object
   savePaymentDetails(paymentInstrument, order, result) {
-    if (result.pspReference) {
-      paymentInstrument.paymentTransaction.transactionID = result.pspReference;
-      order.custom.Adyen_pspReference = result.pspReference;
+    let pspRef;
```
Comment (wboereboom) [inline src/cartridges/int_adyen_overlay/cartridge/scripts/util/adyenHelper.js:554]:
```suggestion
    const pspRef = session.privacy.giftCardResponse ? JSON.parse(session.privacy.giftCardResponse).pspReference: result.pspReference ;
```

File: src/cartridges/int_adyen_overlay/cartridge/scripts/util/adyenHelper.js [hunk 17]
```diff
@@ -565,9 +571,7 @@ var adyenHelperObj = {
         : '';
     order.custom.Adyen_value = '0';
     // Save full response to transaction custom attribute
-    paymentInstrument.paymentTransaction.custom.Adyen_log = JSON.stringify(
-        result,
-    );
+    paymentInstrument.paymentTransaction.custom.Adyen_log =  JSON.stringify(result); //for separate
```
Comment (wboereboom) [inline src/cartridges/int_adyen_overlay/cartridge/scripts/util/adyenHelper.js:574]: I don't understand the comment

**Review by wboereboom — CHANGES_REQUESTED** [review]
There's some files in /cartridges here that do not match the actual code in src it seems.

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/checkout_services/placeOrder.js (hunk 14 above)
Comment (zaiddreakh) [reply src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/checkout_services/placeOrder.js:147]: yes

File: src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderGenericComponent.js (hunk 1 above)
Comment (zaiddreakh) [reply src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderGenericComponent.js:176]: linter wanted a return.. "some" function breaks after satisfying the condition once so the return is not doing any harm

File: src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderGenericComponent.js [hunk 18]
```diff
@@ -135,13 +161,23 @@ module.exports.renderGenericComponent = async function renderGenericComponent()
...
+    imagePath: session.imagePath,
+    adyenDescriptions: session.adyenDescriptions,
   };
   store.checkout = await AdyenCheckout(store.checkoutConfiguration);
   setCheckoutConfiguration(store.checkout.options);
   setInstallments(store.checkout.options.amount);
   setAmazonPayConfig(store.checkout.paymentMethodsResponse);
   document.querySelector('#paymentMethodsList').innerHTML = '';
 
+  // eslint-disable-next-line no-restricted-syntax
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderGenericComponent.js:173]: Can we get rid of the eslint ignore?

File: src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderPaymentMethod.js [hunk 19]
```diff
@@ -133,17 +131,22 @@ function handleInput({ paymentMethodID }) {
     helpers.displaySelectedMethod(event.target.value);
   };
 }
-
+// eslint-disable-next-line complexity
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/renderPaymentMethod.js:134]: Can we get rid of the eslint ignore?

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/Adyen.js [hunk 20]
```diff
@@ -73,6 +73,34 @@ server.post(
...
+ * Called by Adyen to create a partial payments order
+ */
+server.post(
+  'PartialPaymentsOrder',
+  server.middleware.https,
+  adyen.partialPaymentsOrder,
+);
+
+/**
+ * Called by Adyen to make apply a giftcard
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/Adyen.js:100]:
```suggestion
 * Called by Adyen to apply a giftcard
```

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/cancelPartialPaymentOrder.js [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+    const cancelOrderRequest = {
+      merchantAccount: AdyenConfigs.getAdyenMerchantAccount(),
+      order: partialPaymentsOrder,
+    };
+
+    const response = adyenCheckout.doCancelPartialPaymentOrderCall(
+      cancelOrderRequest,
+    );
+
+    if (response.resultCode === 'Received') {
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/cancelPartialPaymentOrder.js:22]: What if we get anything other than received?

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/partialPayment.js [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+function makePartialPayment(req, res, next) {
+  try {
+    const request = JSON.parse(req.body);
+
+    const { paymentMethod, partialPaymentsOrder, amount } = request;
+
+    const partialPaymentRequest = {
+      merchantAccount: AdyenConfigs.getAdyenMerchantAccount(),
+      amount,
+      reference: 'partialPaymentRef',
```
Comment (wboereboom) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/partialPayment.js:17]: Should we find a suitable value for this? maybe something tying to a basket or something else? (I guess we do not have an order at this stage...)

File: src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/cancelPartialPaymentOrder.js [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
```
Comment (maassenbas) [inline src/cartridges/int_adyen_SFRA/cartridge/controllers/middlewares/adyen/cancelPartialPaymentOrder.js:23]: Can also use constant files from backend here

File: src/cartridges/int_adyen_overlay/cartridge/scripts/util/adyenHelper.js [hunk 24]
```diff
@@ -618,6 +625,16 @@ var adyenHelperObj = {
     return format;
   },
 
+  getDivisorForCurrency(amount) {
+    let fractionDigits = adyenHelperObj.getFractionDigits(amount.currencyCode);
+    let divideBy = 1;
```
Comment (maassenbas) [inline src/cartridges/int_adyen_overlay/cartridge/scripts/util/adyenHelper.js:630]: I think we can make this easier, for example we can use `divideBy = Math.pow(10, fractionDigits)`

File: cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/checkoutConfiguration.js [hunk 25]
```diff
@@ -103,6 +109,135 @@ function getGooglePayConfig() {
...
+  $.ajax({
+    type: 'POST',
+    url: 'Adyen-CancelPartialPaymentOrder',
+    data: JSON.stringify(store.partialPaymentsOrderObj),
+    contentType: 'application/json; charset=utf-8',
+    async: false,
+    success: function success(res) {
+      store.partialPaymentsOrderObj = null;
+
+      if (res.resultCode === 'Received') {
```
Comment (amihajlovski) [inline cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/checkoutConfiguration.js:122]: Maybe its good this to be added to a constant file?

Comment (wboereboom) [reply cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/checkoutConfiguration.js:122]: It already is! constants has a `RESULTCODES.RECEIVED` field

Comment (zaiddreakh) [reply cartridges/int_adyen_SFRA/cartridge/client/default/js/adyen_checkout/checkoutConfiguration.js:122]: this is just the autogenerated file, we will make sure it's up to date before we release
This change was already made in the original file inside /src

**Review by wboereboom — APPROVED** [review]

**Review by maassenbas — APPROVED** [review]

**Pull request merged**